### ISSUE TITLE
has_many serializer inference

### ISSUE THREAD

**Issue by ThomasSevestre** [fast-jsonapi/fast_jsonapi#14]
Serializer inference on relationships can be improved.
For instance `belongs_to :owner, record_type: :user` uses `OwnerSerializer`.
In my opinion, it would make more sens to use `UserSerializer`.

I have a patch here: https://github.com/ThomasSevestre/fast_jsonapi/commit/46b5fdefb675159d89ec01b66cb45dea1590cbdf

It is a breaking change, I'll make a PR if you are interested.

**Comment by stas** [comment]
Lets get some more feedback regarding this first.
Personally I do prefer a more explicit way to provide the `type` and the `serializer`, but I don't have a strong opinion.

**Comment by kpheasey** [comment]
I agree, this has been confusing for me a few times.  However, I haven't had the time to come up with a clear set of rules to implement.  @ThomasSevestre, do you have suggestions?

**Comment by ThomasSevestre** [comment]
I beleive it should be the least surprising as possible. IMO the behavior should be :
- use serializer option if specified
- use record_type option if specified
- use relation name

**Comment by christophersansone** [comment]
I'll throw in my opinion since I just created #29 which is somewhat related.

The `record_type` inference and `serializer` inference can both be misleading.  Personally, if I every need to specify one, I specify the other as well.  Here's the thing though: _if it knew which serializer to use, then it should just look up the record type on the serializer_.  To me, the value specified by `set_type` of the serializer is the Single Source of Truth.

Therefore, my opinion is that the `record_type` option is completely superfluous because we offer the `serializer` option.  I think `record_type` should be deprecated in favor of specifying `serializer` directly, and the `type` specified in the output should always come from the serializer.

This way there is only one thing being inferred: detecting the serializer class from the `serializer` option on the relationship.

If we deprecated `record_type`, then this is how I think we can still gracefully support it in the meantime:

* If `record_type` is specified and `serializer` is not specified, simply treat the value of `record_type` as the `serializer` option instead.

* If both `record_type` and `serializer` are specified, ignore `record_type` in favor of the serializer's.  I imagine that in all real-world usage, these match already.  If not, I think the `type` in the `relationships` hash would not match the `type` of the actual serialized record, which would break the JSON:API spec.

What does everyone else think?

**Comment by christophersansone** [comment]
While working on #29, I discovered this in the specs.  This is the JSON output of [this test](https://github.com/fast-jsonapi/fast_jsonapi/blob/1d7c18f5dac0177a0a4d8728d84b4fd63b2d91df/spec/lib/object_serializer_class_methods_spec.rb#L462), shortened to highlight the problem:

```json
{
  "data": [
    {
      "id": "1",
      "type": "movie",
      "attributes": {
        "name": "test movie",
        "release-year": null
      },
      "relationships": {
        "movie-type": {
          "data": {
            "id": "1",
            "type": "movie-type"
          }
        }
      }
    }
  ],
  "included": [
    {
      "id": "1",
      "type": "dash-movie-type",
      "attributes": {
        "name": "Episode"
      }
    }
  ]
}
```

Notice how the value for `type` is different in the `relationships` from in the `included`.  This JSON:API output is invalid because of how it is resolving record types and serializers.

There is even a test to make sure the `type` in the `relationships` equals `movie-type`, presumably because the relationship is defined as `belongs_to :movie_type`.  To me, this test is not correct, and the `type` should always be the type defined on the serializer, in this case `dash-movie-type`.

To me, I think all these problems disappear when we make the serializer the Single Source of Truth, and relationships specify a `serializer` option if it can't be inferred from the relationship name.

**Comment by ThomasSevestre** [comment]
@christophersansone it looks goot to me !
